include/wil/network.h

```diff
@@ -0,0 +1,1328 @@
+#endif
+
+        ~socket_address() = default;
+        socket_address(const socket_address&) WI_NOEXCEPT = default;
+        socket_address& operator=(const socket_address&) WI_NOEXCEPT = default;
+        socket_address(socket_address&&) WI_NOEXCEPT = default;
+        socket_address& operator=(socket_address&&) WI_NOEXCEPT = default;
+
+        bool operator==(const socket_address&) const WI_NOEXCEPT;
+        bool operator!=(const socket_address&) const WI_NOEXCEPT;
+        bool operator<(const socket_address&) const WI_NOEXCEPT;
+        bool operator>(const socket_address&) const WI_NOEXCEPT;
+
+        void swap(socket_address&) WI_NOEXCEPT;
+        void reset(ADDRESS_FAMILY family = AF_UNSPEC) WI_NOEXCEPT;
+
+        // set_sockaddr overwrites the entire sockaddr in the object (including address family)
+        template <typename T>
+        void set_sockaddr(_In_reads_bytes_(addr_size) const SOCKADDR* addr, T addr_size) WI_NOEXCEPT;
+        void set_sockaddr(const SOCKADDR_IN*) WI_NOEXCEPT;
+        void set_sockaddr(const SOCKADDR_IN6*) WI_NOEXCEPT;
+        void set_sockaddr(const SOCKADDR_INET*) WI_NOEXCEPT;
+        void set_sockaddr(const SOCKET_ADDRESS*) WI_NOEXCEPT;
+#if defined(WIL_ENABLE_EXCEPTIONS)
+        void set_sockaddr(SOCKET);
+        void set_sockaddr(PCWSTR);
+        void set_sockaddr(PCSTR);
+#endif
+        [[nodiscard]] HRESULT set_sockaddr_nothrow(SOCKET) WI_NOEXCEPT;
+        [[nodiscard]] HRESULT set_sockaddr_nothrow(PCWSTR) WI_NOEXCEPT;
+        [[nodiscard]] HRESULT set_sockaddr_nothrow(PCSTR) WI_NOEXCEPT;
+
+        // set_address* preserves the existing port set on the address
+        // - so that one can call set_port() and set_address() in any order with expected results
+        // the address family is preserved unless it is specified (or inferred) as an argument
+        void set_address_any() WI_NOEXCEPT;
+        void set_address_any(ADDRESS_FAMILY family) WI_NOEXCEPT;
+        void set_address_loopback() WI_NOEXCEPT;
+        void set_address_loopback(ADDRESS_FAMILY family) WI_NOEXCEPT;
+        void set_address(const IN_ADDR*) WI_NOEXCEPT;
+        void set_address(const IN6_ADDR*) WI_NOEXCEPT;
+
+        void set_port(USHORT) WI_NOEXCEPT;
+        void set_scope_id(ULONG) WI_NOEXCEPT;
+        void set_flow_info(ULONG) WI_NOEXCEPT;
+
+        // write_address prints the IP address, not the scope id or port
+        HRESULT write_address_nothrow(socket_address_wstring& address) const WI_NOEXCEPT;
+        HRESULT write_address_nothrow(socket_address_string& address) const WI_NOEXCEPT;
+
+        // write_complete_address_nothrow() prints the IP address as well as the scope id and port values
+        HRESULT write_complete_address_nothrow(socket_address_wstring& address) const WI_NOEXCEPT;
+        HRESULT write_complete_address_nothrow(socket_address_string& address) const WI_NOEXCEPT;
+
+#if defined(WIL_ENABLE_EXCEPTIONS) && (defined(_STRING_) || defined(WIL_DOXYGEN))

```

Comment:
Don't do this (checking for the definition of `_STRING_` etc.). See #490 for examples on how to use the STL in WIL moving forward